File: June/Week 2/Insert Delete GetRandomO(1).go
import (
	"fmt"
	"math/rand"
)

type RandomizedSet struct {
	values map[int]int
	arr    []int
}

/** Initialize your data structure here. */
func Constructor() RandomizedSet {
	return RandomizedSet{make(map[int]int), make([]int, 0, 100)}
}

/** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
func (this *RandomizedSet) Insert(val int) bool {
	_, exists := this.values[val]
	if exists {
		return false
	}
	this.values[val] = len(this.arr)
	this.arr = append(this.arr, val)
	return true
}

/** Removes a value from the set. Returns true if the set contained the specified element. */
func (this *RandomizedSet) Remove(val int) bool {
	currIndex, exists := this.values[val]
	if !exists {
		return false
	}
	lastEle := this.arr[len(this.arr)-1]
	this.arr[currIndex] = this.arr[len(this.arr)-1]
	this.arr = this.arr[:(len(this.arr) - 1)]
	this.values[lastEle] = currIndex
	delete(this.values, val)
	return true
}

/** Get a random element from the set. */
func (this *RandomizedSet) GetRandom() int {
	temp := rand.Intn((len(this.arr) - 1) | 1)
	fmt.Println(temp)
	return this.arr[temp]
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Insert(val);
 * param_2 := obj.Remove(val);
 * param_3 := obj.GetRandom();
 */

File: June/Week 2/Insert Delete GetRandomO(1).ts
class RandomizedSet {
    indexMap: Map<number, number>;
    arr: number[];
    constructor() {
        this.indexMap = new Map<number, number>();
        this.arr = [];
    }

    insert(val: number): boolean {
        if (this.indexMap.has(val)) {
            return false;
        }
        this.indexMap.set(val, this.arr.length);
        this.arr.push(val)
        return true;
    }

    remove(val: number): boolean {
        if (!this.indexMap.has(val)) {
            return false;
        }
        let currIndex: number = this.indexMap.get(val) || 0;
        this.arr[currIndex] = this.arr[this.arr.length - 1];
        this.indexMap.set(this.arr[currIndex], currIndex);
        this.indexMap.delete(val);
        this.arr.pop();
        return true;
    }

    getRandom(): number {
        // let temp = Math.floor(Math.random()*this.arr.length);
        return this.arr[Math.floor(Math.random() * this.arr.length)];
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * var obj = new RandomizedSet()
 * var param_1 = obj.insert(val)
 * var param_2 = obj.remove(val)
 * var param_3 = obj.getRandom()
 */